libslide/src/intern.rs

```diff
@@ -0,0 +1,62 @@
+//! Interns items in a common repository.
+
+use core::ops::Deref;
+use std::collections::HashMap;
+use std::fmt::Display;
+use std::hash::Hash;
+
+#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub struct Interned<'intern, T> {
+    inner: &'intern T,
+}
+
+impl<'intern, T> Deref for Interned<'intern, T>
+where
+    T: Hash + Eq + Clone + Display,
+{
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<'intern, T> std::fmt::Display for Interned<'intern, T>
+where
+    T: Hash + Eq + Clone + Display,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.inner.fmt(f)
+    }
+}
+
+pub struct Interner<'intern, T>
+where
+    T: Hash + Eq + Clone + Display,
+{
+    interned: HashMap<T, &'intern T>,
+}
+
+impl<'intern, T> Default for Interner<'intern, T>
+where
+    T: Hash + Eq + Clone + Display,
+{
+    fn default() -> Self {
+        Self {
+            interned: HashMap::new(),
+        }
+    }
+}
+
+impl<'intern, T> Interner<'intern, T>
+where
+    T: Hash + Eq + Clone + Display,
+{
+    pub fn intern(&mut self, val: T) -> Interned<T> {
+        let inner = match self.interned.get(&val) {
+            Some(interned) => interned,
+            None => self.interned.entry(val.clone()).or_insert(val),
+        };
+        Interned { inner }
+    }
+}
```

Comment:
Maybe, I dont understand interners, but do we need both an interned, and intern, and an interner classes here? Seems like an over complication